### PR TITLE
ar71xx, ath79: fix MAC address setup for TL-WDR4300 board

### PR DESCRIPTION
**Pull request by mans0n**
The current ethernet MAC address setup of TL-WDR4300 board is different
from the setup of stock firmware:
```
OpenWrt: lan = label_mac -2, wan = label_mac -2
  stock: lan = label_mac,    wan = label_mac +1
```
This patch applies to all devices using TL-WDR4300 board:
TL-WDR3600 v1
TL-WDR4300 v1
TL-WDR4300 v1 (IL)
TL-WDR4310 v1
Mercury MW4530R v1

Ref: #2600

### CODE REVIEW

**Comment by mans0n** [comment]
I wanted to change setup of TL-WDR4300 only, but (on ar71xx) those devices use the same model name and it seems difficult to split them. So this should be verified on the other devices too.

**Comment by Leo-PL** [comment]
I think it's not to worry about breaking others really, WDR3600 is literally WDR4300 without one 5GHz radio chain. Too bad I haven't got one to play with, WDR4300 is my "trusty hardware".

File: target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration [hunk 1]
```diff
@@ -29,6 +29,11 @@ tplink,re355-v1)
 tplink,re450-v1)
 	migrate_leds "re450:=tp-link:"
 	;;
+tplink,tl-wdr3600-v1|\
+tplink,tl-wdr4300-v1|\
+tplink,tl-wdr4300-v1-il)
+	migrate_leds ":blue:=:green:"
```
Comment (Leo-PL) [inline target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration]: It's great you spotted this, +1.

**Review by Leo-PL — APPROVED** [review]

**Comment by vlad3055** [comment]
So you're proposing this:?
```
wlan2g: label_mac -1
wlan5g: label_mac   ***duplicate
lan: label_mac   ***duplicate
wan: label_mac +1
```

I'm against duplicate mac addresses, it can cause ARP leaks

If we're opening the issue of mac addresses, why not apply a different mac address to each of the 4 interface?

on my wdr4300 the sticker says: 14:cc:20:xx:xx:23
the last 2 bits are on.

I propose this:
```
wlan2g: label_mac
wlan5g: label_mac -1
lan: label_mac -2
wan: label_mac -3
```

**Comment by mans0n** [comment]
@vlad3055 AFAIK using same MAC address between wired and wireless interfaces(e.g. eth0.1 and wlan0) is quite common case. And I'm just following stock firmware's behavior, so there won't be any issue if stock firmware already works without a problem. I'm more against using different address from stock firmware, as it might cause (as you said) ARP leaks over stock device and OpenWrt device.(e.g. LAN address of :EE OpenWrt device might be same as LAN address of :EC stock device, causing collision)

**Comment by vlad3055** [comment]
so the plot thickens...:

I've tested the greatest common divisor (gcd) of the difference of mac labels of 5 different wdr4300's I have lying around. It's 1.

gcd(mac2-mac1, mac3-mac2, mac4-mac3, mac5-mac4) == 1

so it appears tplink don't care too much about collisions between WAN of one router to LAN of next router.

The stock firmware is rather primitive compared to the openwrt, I still think we should use 4 different MACs for openwrt to avoid internal mac collisions and arp leaks inside a single openwrt router

**Comment by mans0n** [comment]
@vlad3055 Hmm...I guess you have a point. We would need a verdict from the core developers...

**Comment by Leo-PL** [comment]
Also, guys, please differentiate MACs of interfaces from default BSSIDs. Even if BSSID is different, the router's bridge interface will appear with same MAC from both wireless interfaces and from wired one.

File: target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration (hunk 1 above)
Comment (Leo-PL) [reply target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration]: Also, perhaps you could merge that into archer-c7-v2 case.

**Comment by vlad3055** [comment]
I agree, they all need to be different. The bssid's are labeled "wlan2g/5g" below

In stock firmware it doesn't matter that wlan5g==lan because there is a single bridge, but in openwrt wlan5g may be bridged to another interface/lan/vpn instead.

btw that duplicate mac-address problem may appear also in other openwrt configurations, not just wdr4300...

```
wlan2g(bssid): label_mac
wlan5g(bssid): label_mac -1
lan(mac): label_mac -2
wan(mac): label_mac -3
```

**Comment by adrianschmutzler** [comment]
It's OpenWrt policy to stick with the vendor's assignment of MAC addresses (unless for the very rare cases where they have random addresses on particular interfaces). When assigning addresses by personal taste, you both have the problem of different taste causing collisions and less predictability for the user, as you just increase randomness in the whole code base.

In particular, note that devices having distinct addresses for all interfaces on vendor FW are actually relatively uncommon. So, NAK on personal choice addresses.

**Review by Leo-PL — CHANGES_REQUESTED** [review]
LED migration has merit, but probably this should be merged with tplink,archer-c7-v2 case, avoiding code duplication.

File: target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration (hunk 1 above)
Comment (mans0n) [reply target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration]: Oh, I didn't noticed that. Merged.

**Comment by mans0n** [comment]
Merged duplicate cases.

**Comment by adrianschmutzler** [comment]
I've pulled the LED migration to my staging tree:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

I will wait with the MAC address setup until I have time to verify on my device.

I will backport to 19.07 and 18.06 as necessary after a short waiting period.

**Comment by Leo-PL** [comment]
@adrianschmutzler Regarding extended commit message about LED colors - AFAICT up to V1.5 blue LEDs were used, V1.6 revision changed all front LEDs to green. Rear USB LEDs were always green, at least since V1.2.

**Comment by adrianschmutzler** [comment]
> Regarding extended commit message about LED colors - AFAICT up to V1.5 blue LEDs were used, V1.6 revision changed all front LEDs to green. Rear USB LEDs were always green, at least since V1.2.

Thanks, but I don't think finding out the precise assignment will add value to the commit. I just added the general comment to prevent people from changing it back and forth in the future, as they find their device to be different ...

I only changed it in a similar case for WDR4900v2 because there the setup for ar71xx was actually broken.

**Comment by Leo-PL** [comment]
OK, just wanted to point out the reason is a bit different, my first-hand information isn't that precise either, I haven't had V1.5 in my hands ;)

**Comment by adrianschmutzler** [comment]
Ah, so you mean I should change "different by region" to "different by version" ...

**Comment by adrianschmutzler** [comment]
I might receive information about a WDR3600 during the week ...

**Comment by Leo-PL** [comment]
> Ah, so you mean I should change "different by region" to "different by version" ...

Yes, exactly.

**Comment by adrianschmutzler** [comment]
WDR3600 stock:
LAN label
2G label-1
5G label
WAN label+1

**Comment by adrianschmutzler** [comment]
Merged and backported right away.

**Pull request closed without merging**